rfcs/20240315-algorithm-support.md

```diff
@@ -0,0 +1,166 @@
+# [RFC] Add algorithm to dot_general's parameters in the StableHLO specification
+
+Status: Review<br/>
+Initial version: 03/15/2024<br/>
+Last updated: 03/15/2024<br/>
+Discussion thread: [GitHub](https://github.com/openxla/stablehlo/pull/2096)
+
+## Motivation
+
+We would like to allow explicitly selecting the algorithm used for individual
+`dot_general` instructions, enforcing a well-defined numeric precision. We think
+that the current precision values (`DEFAULT`, `HIGH`, `HIGHEST`) are not
+specific enough for this, because they mean different things between hardware
+platforms and even between GPU types.
+
+We also want to make the API flexible enough to support additional algorithms
+that might be faster on certain hardware
+([6xBF16](https://arxiv.org/pdf/1904.06376.pdf), 3xTF32, etc) that can't fit
+into the limited precision_config values of `DEFAULT`, `HIGH` and `HIGHEST`. For
+example "6xBF16" provides similar precision to F32, but about 2x performance on
+some GPUs.
+
+When an algorithm is not supported by an accelerator, the program should fail
+instead of implicitly falling back to another. It should be the responsibility
+of higher level frameworks such as JAX, to define the required algorithm for
+each accelerator that is used.
+
+## Proposed Specification change
+
+### dot_general
+
+#### Semantics
+
+*Add these lines:*
+
+`algorithm` defines the main properties of the algorithm used to implement the
+dot operation, which also defines the precision. If the algorithm value is set
+(`!= UNSET`), then the `precision_config` is ignored. If the algorithm value is
+not set (`= UNSET`), then the `precision_config` will take effect.
+
+In general the first two types in the algorithm name are `AType` and `BType` -
+the precisions that the LHS and RHS of the operation are rounded to, and the
+third one is `AccumType` - the accumulation type. These types are independent
+from the storage types of the inputs and the output (but not all combinations
+are supported). Some algorithms have other suffixes too.
+
+This is the current list of algorithms:
+
+* `UNSET`: There is no restriction on the algorithm, the precision is decided
+  based on the `precision_config`. Use this for every type combination that
+  doesn't have a specific algorithm.
+* `ANY_F8_ANY_F8_F32`: The inputs are used with the full precision of their
+  storage type (which must be an 8-bit floating point type - the same for both
```

Comment:
Why must a "full precision" be 8bit floating point type? Do you assume that this algorithm is only applicable to fp8 matmul? What happens if someone use this algorithm for a fp32 matmul?

Reply:
Well ANY_F8_ANY_F8_F32 was defined only for F8 types, but actually we may rather use structs instead of enums as described in https://github.com/openxla/stablehlo/pull/2096#discussion_r1616185205, so this problem will not come up.